### PR TITLE
EDD-26: Adds file list header, refactors download list and files list

### PR DESCRIPTION
**Pull request by macrouch**
# Overview

### What is the feature?

Adds a file list header. 
Refactored download list and file list to both utilize the `ListPage` component. 
Implemented react-window and refactored progress reporting to address performance issues in files list

### What areas of the application does this impact?

Download list, files list

# Testing

### Reproduction steps

Start a new download, ensure the download list still works as expected.
Click the download to view the files list
* Updates should be displayed every second without any performance slow downs.
* Files header shows the information related to the download
* Files header actions work as expected

# Checklist

- [x] I have added automated tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have bumped the `version` field in package.json and ran `npm install`

### CODE REVIEW

File: src/main/eventHandlers/requestDownloadsProgress.ts [hunk 1]
```diff
@@ -8,27 +8,29 @@ import downloadStates from '../../app/constants/downloadStates'
...
+  } = info
 
-  if (downloads.length === 0) {
-    webContents.send('reportDownloadsProgress', {
-      progressReport: []
-    })
+  const totalDownloads = await database.getAllDownloadsCount()
 
-    return
+  if (totalDownloads.length === 0) {
```
Comment (macrouch) [inline src/main/eventHandlers/requestDownloadsProgress.ts:22]: this shouldn't be .length

File: src/main/utils/database/EddDatabase.ts [hunk 2]
```diff
@@ -229,34 +276,17 @@ class EddDatabase {
...
+      .update(data)
+      .where(where)
   }
 
   /**
    * Returns the files progress for the given downloadId
    * @param {String} downloadId ID of download to return progress
    */
-  async getDownloadFilesProgressByDownloadId(downloadId) {
+  // async getDownloadFilesProgressByDownloadId(downloadId) {
```
Comment (macrouch) [inline src/main/utils/database/EddDatabase.ts:288]: Remove

File: src/main/utils/database/EddDatabase.ts [hunk 3]
```diff
@@ -276,6 +306,151 @@ class EddDatabase {
...
+
+  /**
+   * Returns the data needed for the Downloads report
+   * @param {Number} limit Limit of the files returned
+   * @param {Number} offset Offset of the files returned
+   */
+  async getDownloadsReport(limit, offset) {
+    return this.db('downloads')
+      .select(
+        // 'downloads.*',
```
Comment (macrouch) [inline src/main/utils/database/EddDatabase.ts:430]: remove

File: src/app/main.jsx [hunk 4]
```diff
@@ -3,7 +3,10 @@ import ReactDOM from 'react-dom/client'
 import App from './App'
 
 ReactDOM.createRoot(document.getElementById('root')).render(
-  <React.StrictMode>
-    <App />
-  </React.StrictMode>
+  <App />
 )
+// ReactDOM.createRoot(document.getElementById('root')).render(
```
Comment (macrouch) [inline src/app/main.jsx:8]: revert change

File: src/app/utils/parseDownloadReport.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
+import downloadStates from '../constants/downloadStates'
+
+/**
+ * Parses the DownloadReport into values used by components
+ * @param {Object} report Report generated by the main process for the download
+ */
+const parseDownloadReport = (report) => {
```
Comment (eudoroolivares2016) [inline src/app/utils/parseDownloadReport.js:7]: Getting a console error if I start up the application without any downloads can you confirm if you have same error in log? `ncaught (in promise) TypeError: Cannot read properties of undefined (reading 'length')
    at parseDownloadReport (parseDownloadReport.js:10:40)
    at loadItems (Downloads.jsx:114:28)`

Comment (macrouch) [reply src/app/utils/parseDownloadReport.js:7]: Yep I had seen that, just pushed the fix

File: src/app/components/DownloadHeader/__tests__/DownloadHeader.test.js [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+      })
+
+      expect(screen.getByText('Paused')).toHaveClass('derivedStatus')
+      expect(screen.getByText('5 of 10 files done')).toHaveClass('humanizedStatus')
+
+      expect(screen.queryByRole('button', { name: 'Pause All' })).not.toBeInTheDocument()
+      expect(screen.getByRole('button', { name: 'Resume All' })).toBeInTheDocument()
+      expect(screen.getByRole('button', { name: 'Cancel All' })).toBeInTheDocument()
+    })
+  })
```
Comment (eudoroolivares2016) [inline src/app/components/DownloadHeader/__tests__/DownloadHeader.test.js:141]: Does it make sense to add a test for if all the file downloads are cancelled? Since none of the buttons are rendering so something like:

  describe('when all downloads are cancelled', () => {
    test('displays the correct information', () => {
      setup({
        allDownloadsPaused: false,
        state: downloadStates.cancelled,
        totalCompletedFiles: 0,
        totalFiles: 10
      })

      expect(screen.getByText('Paused')).toHaveClass('derivedStatus')
      expect(screen.getByText('5 of 10 files done')).toHaveClass('humanizedStatus')

      expect(screen.queryByRole('button', { name: 'Pause All' })).not.toBeInTheDocument()
      expect(screen.getByRole('button', { name: 'Resume All' })).not.toBeInTheDocument()
      expect(screen.getByRole('button', { name: 'Cancel All' })).not.toBeInTheDocument()
    })

Comment (macrouch) [reply src/app/components/DownloadHeader/__tests__/DownloadHeader.test.js:141]: There are no checks in the component for state=cancelled. When you cancel a running component it goes to completed. So there is no way to write this test

File: src/app/components/DownloadListItem/DownloadListItem.jsx [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+import DownloadListItemState from './DownloadListItemState'
+import DownloadListItemPercent from './DownloadListItemPercent'
+import DownloadListItemFileProgress from './DownloadListItemFileProgress'
+import { ElectronApiContext } from '../../context/ElectronApiContext'
+import useAppContext from '../../hooks/useAppContext'
+
+/**
+ * @typedef {Object} DownloadListItemProps
+ * @property {Object} download State of the download item.
+ * @property {Function} setCurrentPage A function which sets the active page.
```
Comment (eudoroolivares2016) [inline src/app/components/DownloadListItem/DownloadListItem.jsx:28]:
```suggestion
 * @property {Function} setCurrentPage A function which sets the active page.
 * @property {Function} setSelectedDownloadId A function which sets the setSelectedDownloadI.
```

File: src/app/components/FileDownloadsHeader/FileDownloadsHeader.jsx [hunk 8]
```diff
@@ -1,23 +1,38 @@
...
+import * as styles from './FileDownloadsHeader.module.scss'
 
-// TODO EDD-26 add styling to component
 /**
  * @typedef {Object} FileDownloadsHeaderProps
- * @property {Boolean} checked State of the hide completed checkbox.
+ * @property {Boolean} hideCompleted State of the hide completed checkbox.
  * @property {Function} setCheckboxState A function to update the state of the  `hide completed` checkbox.
- * @property {Function} setCurrentPage A function which sets the active page.
+ * @property {Function} setHideCompleted A function which sets the active page.
```
Comment (eudoroolivares2016) [inline src/app/components/FileDownloadsHeader/FileDownloadsHeader.jsx:35]:
```suggestion
 * @property {Function} setHideCompleted A function which sets `hideCompleted` downloads state.
 * @property {Function} setCurrentPage A function which sets the active page.
```

File: src/app/components/FileDownloadsHeader/FileDownloadsHeader.jsx [hunk 9]
```diff
@@ -49,112 +63,237 @@ const FileDownloadsHeader = ({
...
+    <div className={styles.wrapper}>
+      <div className={styles.header}>
+        <Button
+          className={styles.backLink}
+          Icon={FaChevronLeft}
+          size="lg"
+          onClick={
+            () => {
+              // We start interval to poll the state of the downloads
+              startReportingDownloads()
```
Comment (eudoroolivares2016) [inline src/app/components/FileDownloadsHeader/FileDownloadsHeader.jsx:129]: Do we still need this function? `startReportingDownloads()` I removed it locally and I was still able to get download updates when I went back to downloads page

Comment (eudoroolivares2016) [reply src/app/components/FileDownloadsHeader/FileDownloadsHeader.jsx:129]: I removed all references from the main ipc functions `startReportingFiles: (data) => ipcRenderer.send('startReportingFiles', data),
 startReportingDownloads: () => ipcRenderer.send('startReportingDownloads'),
`
The diff also removed the file definitions for both of these

File: src/app/components/FileDownloadsHeader/__tests__/FileDownloadsHeader.test.js [hunk 10]
```diff
@@ -46,36 +58,251 @@ const setup = (overrideProps) => {
...
+
+      const button = screen.getByRole('checkbox', { name: 'Hide Completed' })
+      await userEvent.click(button)
+
+      expect(setHideCompleted).toHaveBeenCalledTimes(1)
+      expect(setHideCompleted).toHaveBeenCalledWith(true)
+    })
+  })
+
+  describe('when the download is pending', () => {
```
Comment (eudoroolivares2016) [inline src/app/components/FileDownloadsHeader/__tests__/FileDownloadsHeader.test.js:153]: Not a PR comment for the test but, for my own understanding, are we able to get to this view on the interface? Or does disabling the fileDownload from being clickable prevent that? (this is what I think I'm seeing)

File: src/app/components/ListPage/ListPage.jsx [hunk 11]
```diff
@@ -87,12 +106,43 @@ const ListPage = ({
...
+                                itemCount={items.length}
+                                itemData={items}
+                                itemSize={97}
+                                onItemsRendered={onItemsRendered}
+                                outerRef={scrollableNodeRef}
+                                ref={listRef}
+                                width={width}
+                                overscanCount={4}
+                              >
+                                {ListPageListItem}
```
Comment (eudoroolivares2016) [inline src/app/components/ListPage/ListPage.jsx:139]: I'm a bit confused about how `ListPageListItem` is getting its props set since they aren't being passed in here?

File: src/app/pages/Downloads/Downloads.jsx [hunk 12]
```diff
@@ -1,45 +1,28 @@
...
 
 import * as styles from './Downloads.module.scss'
+
 /**
  * @typedef {Object} DownloadsProps
- * @property {Boolean|Number} hasActiveDownload A boolean for whether there is an active download taking place.
  * @property {Function} setCurrentPage A function which sets the active page.
- * @property {Function} showMoreInfoDialog A function to set the `showMoreInfoDialog` in the layout
+ * @property {Function} showMoreInfoDialog A function to set the `showMoreInfoDialog` in the layout.
  * @property {Function} setHasActiveDownload A function to set whether a download is active.
```
Comment (eudoroolivares2016) [inline src/app/pages/Downloads/Downloads.jsx:26]:
```suggestion
 * @property {Function} setHasActiveDownload A function to set whether a download is active.
 * @property {Function} setSelectedDownloadId A function which sets the setSelectedDownloadId.
```

File: src/app/pages/Downloads/Downloads.jsx [hunk 13]
```diff
@@ -50,309 +33,115 @@ import * as styles from './Downloads.module.scss'
  *
  * return (
  *   <Downloads
- *     hasActiveDownload={hasActiveDownload}
  *     setCurrentPage={setCurrentPage}
  *     showMoreInfoDialog={showMoreInfoDialog}
  *     setHasActiveDownload={setHasActiveDownload}
```
Comment (eudoroolivares2016) [inline src/app/pages/Downloads/Downloads.jsx:38]:
```suggestion
 *     setHasActiveDownload={setHasActiveDownload}
 *     setSelectedDownloadId={setSelectedDownloadId}
```

File: src/app/pages/Downloads/Downloads.jsx [hunk 14]
```diff
@@ -50,309 +33,115 @@ import * as styles from './Downloads.module.scss'
...
+  const listRef = useRef()
+
+  const [items, setItems] = useState([])
+  const [windowState, setWindowState] = useState({})
   const [allDownloadsPaused, setAllDownloadsPaused] = useState(false)
   const [allDownloadsCompleted, setAllDownloadsCompleted] = useState(false)
-  const [hasPausedDownload, setHasPausedDownload] = useState(false)
-  const [totalDownloadFiles, setTotalDownloadFiles] = useState(0)
+  const [totalFiles, setTotalFiles] = useState(0)
+  const [totalDownloads, setTotalDownloads] = useState(null)
```
Comment (eudoroolivares2016) [inline src/app/pages/Downloads/Downloads.jsx:59]: Any reason why the  totalDownloads default state wouldn't also be 0?

File: src/main/preload.ts [hunk 15]
```diff
@@ -38,9 +38,9 @@ contextBridge.exposeInMainWorld('electronApi', {
...
 
-  // Messages polled at intervals
-  reportDownloadsProgress: (on, callback) => ipcRenderer[on ? 'on' : 'off']('reportDownloadsProgress', callback),
-  reportFilesProgress: (on, callback) => ipcRenderer[on ? 'on' : 'off']('reportFilesProgress', callback),
+  // Reporting
+  requestDownloadsProgress: (data) => ipcRenderer.invoke('requestDownloadsProgress', data),
+  requestFilesProgress: (data) => ipcRenderer.invoke('requestFilesProgress', data),
 
   // Start functions to report files and downloads and kickoff interval polling
   startReportingFiles: (data) => ipcRenderer.send('startReportingFiles', data),
```
Comment (eudoroolivares2016) [inline src/main/preload.ts:46]: I think these below can be removed

**Review by eudoroolivares2016 — APPROVED** [review]

File: src/app/components/FileListItem/__tests__/FileListItem.test.js [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+}
+
+const setup = (overrideProps = {}) => {
+  const cancelDownloadItem = jest.fn()
+  const copyDownloadPath = jest.fn()
+  const openDownloadFolder = jest.fn()
+  const restartDownload = jest.fn()
+  const showWaitingForEulaDialog = jest.fn()
+  const showWaitingForLoginDialog = jest.fn()
+  const startReportingFiles = jest.fn()
```
Comment (eudoroolivares2016) [inline src/app/components/FileListItem/__tests__/FileListItem.test.js:35]:
```suggestion
  const startReportingFiles = jest.fn()
```
We can remove this mocked function its not called in `fileListItem`

File: src/app/components/ListPage/ListPage.jsx (hunk 11 above)
Comment (macrouch) [reply src/app/components/ListPage/ListPage.jsx:139]: Its just how the react-window library works, https://react-window.vercel.app/#/examples/list/fixed-size

Comment (eudoroolivares2016) [reply src/app/components/ListPage/ListPage.jsx:139]: Thank you!

**Review by trevorlang — APPROVED** [review]
